File: apps/api/src/routes/product-flow.ts
import { Router } from "express";
import type { Request, Response } from "express";
import { config } from "@silo/engine/config";
import { respondServiceError as respondProductServiceError } from "../lib/respond-service-error.js";
import * as productService from "../services/product-service.js";

const router = Router();

function respondProductFlowError(res: Response, status: number, error?: string) {
  res.status(status).json({ success: false, error: error ?? "Erro ao processar fluxo de produto." });
}

function respondProductFlowSuccess(res: Response, entry: unknown) {
  res.status(200).json({ success: true, ok: true, data: { entry }, entry });
}

// POST /api/product-flow/receive
router.post("/receive", async (req: Request, res: Response) => {
  try {
    const apiKeyHeader = req.headers["x-api-key"];
    const providedApiKey = Array.isArray(apiKeyHeader) ? apiKeyHeader[0] : apiKeyHeader;
    const expectedKey = config.productFlowApiKey;
    if (expectedKey && providedApiKey !== expectedKey) {
      respondProductFlowError(res, 401, "Não autorizado.");
      return;
    }

    const body = req.body as { productId?: string; slug?: string; payload?: unknown };
    const { productId, slug, payload } = body ?? {};

    if (!productId && !slug) {
      respondProductFlowError(res, 400, "productId ou slug são obrigatórios no corpo da requisição.");
      return;
    }

    const result = await productService.appendProductFlowEntry({ productId, slug, payload });
    if (!result.ok) {
      respondProductServiceError(res, result, "Erro ao processar fluxo de produto.");
      return;
    }

    respondProductFlowSuccess(res, result.data.entry);
  } catch (error) {
    console.error("[PRODUCT_FLOW] Error receiving flow payload:", error);
    respondProductFlowError(res, 500, "Erro interno do servidor.");
  }
});

export default router;

File: apps/backend/src/silo/api/routers/product_flow.py
from __future__ import annotations

from datetime import UTC, datetime

from fastapi import APIRouter, Depends, Header
from sqlalchemy import select
from sqlalchemy.engine import Connection

from silo.api.dependencies import get_db
from silo.api.responses import json_error_response
from silo.config import load_settings
from silo.db.models import legacy_tables

router = APIRouter(prefix="/api/product-flow", tags=["product-flow"])


@router.post("/receive")
async def receive_product_flow(
    payload: dict[str, object],
    x_api_key: str | None = Header(default=None, alias="x-api-key"),
    db: Connection = Depends(get_db),
):
    settings = load_settings()
    expected_key = settings.product_flow_api_key.get_secret_value()
    if expected_key and x_api_key != expected_key:
        return json_error_response(401, "Não autorizado.")

    product_id = _optional_str(payload.get("productId"))
    slug = _optional_str(payload.get("slug"))
    entry_payload = payload.get("payload")
    if not product_id and not slug:
        return json_error_response(
            400,
            "productId ou slug são obrigatórios no corpo da requisição.",
        )

    product_table = legacy_tables["product"]
    statement = select(product_table.c.id, product_table.c.slug, product_table.c.data_product_flow)
    if product_id:
        statement = statement.where(product_table.c.id == product_id)
    else:
        statement = statement.where(product_table.c.slug == slug)
    product_row = db.execute(statement.limit(1)).mappings().first()
    if product_row is None:
        return json_error_response(404, "Produto não encontrado.")

    entry = {
        "receivedAt": datetime.now(UTC).isoformat().replace("+00:00", "Z"),
        "payload": entry_payload,
    }

    current_flow = product_row["data_product_flow"]
    if not isinstance(current_flow, list):
        current_flow = []
    current_flow = [*current_flow, entry]

    db.execute(
        product_table.update()
        .where(product_table.c.id == product_row["id"])
        .values(data_product_flow=current_flow)
    )
    db.commit()

    return {
        "success": True,
        "ok": True,
        "data": {"entry": entry},
        "entry": entry,
    }


def _optional_str(value: object | None) -> str | None:
    return value if isinstance(value, str) and value.strip() else None
